src/main/java/org/openstreetmap/atlas/geography/atlas/items/Relation.java

```diff
@@ -454,13 +454,13 @@ else if (entity.intersects(surface))
      * {@link PackedAtlas} but could happen when two {@link Atlas} are combined into a
      * {@link MultiAtlas}.
      *
-     * @param polygon
-     *            The {@link Polygon} to check for
+     * @param surface
+     *            The {@link GeometricSurface} to check for
      * @param parentRelationIdentifiers
      *            The identifiers of the parent relations that have already been visited.
      * @return {@code true} if the relation has all members within the given {@link Polygon}
      */
-    protected boolean withinInternal(final Polygon polygon,
+    protected boolean withinInternal(final GeometricSurface surface,
```

Comment:
@jklamer - seems like the [Congestive Complexity is 17](https://sonarcloud.io/project/issues?id=org.openstreetmap.atlas%3Aatlas&issues=AWimF3qKr-MUYPUzxCj4&open=AWimF3qKr-MUYPUzxCj4&pullRequest=334).

Based on what I [read](https://blog.sonarsource.com/cognitive-complexity-because-testability-understandability), the below suggested refactoring should help bring it to below 15 and let the check pass in Sonar.

Hope this helps.

```suggestion
    protected boolean withinInternal(final GeometricSurface surface,
            final Set<Long> parentRelationIdentifiers)
    {
        for (final RelationMember member : this)
        {
            final AtlasEntity entity = member.getEntity();
            if (entity instanceof Relation)
            {
                final long identifier = entity.getIdentifier();
                if (parentRelationIdentifiers.contains(identifier))
                {
                    continue;
                }
                else
                {
                    parentRelationIdentifiers.add(identifier);
                    if (!((Relation) entity).withinInternal(surface, parentRelationIdentifiers))
                    {
                        return false;
                    }
                }
            }
            else if (isUnenclosedNonRelationEntity(surface, entity))
            {
                return false;
            }
        }
        return true;
    }

    private boolean isUnenclosedNonRelationEntity(final GeometricSurface surface,
            final AtlasEntity entity)
    {
        return isUnenclosedLineItem(entity, surface) || isUnenclosedLocationItem(entity, surface)
                || isUnenclosedArea(entity, surface);
    }

    private boolean isUnenclosedArea(final AtlasEntity entity, final GeometricSurface surface)
    {
        return entity instanceof Area
                && !surface.fullyGeometricallyEncloses(((Area) entity).asPolygon());
    }

    private boolean isUnenclosedLocationItem(final AtlasEntity entity,
            final GeometricSurface surface)
    {
        return entity instanceof LocationItem
                && !surface.fullyGeometricallyEncloses(((LocationItem) entity).getLocation());
    }

    private boolean isUnenclosedLineItem(final AtlasEntity entity, final GeometricSurface surface)
    {
        return entity instanceof LineItem
                && !surface.fullyGeometricallyEncloses(((LineItem) entity).asPolyLine());
    }
```